File: gato.cs
using System;

public interface Comida {

	public int Nutrientes();
	public bool ContieneCacao();
}

class Chocolate : Comida{
	public int Nutrientes() {
		return 5;
	}
	
	public bool ContieneCacao() {
		return true;
	}
}

public abstract class Animal{
	private int hambre;
	
	virtual public void Comer(Comida comida){
		if(!this.PuedeComer(comida)){
			throw new Exception("No puede comer la comida");
		}
		
		hambre -= comida.Nutrientes();
	}
		   
	abstract protected bool PuedeComer(Comida comida);
}

public class Gato: Animal{
	
	override public void Comer(Comida comida){
		base.Comer(comida);
		this.DecirMeow();
	}
	
	override protected bool PuedeComer(Comida comida){
		return !comida.ContieneCacao();
	}
	
	private void DecirMeow(){
		Console.Write("Meow");
	
	}

}

File: gato.java
interface Comida {
	public int nutrientes();
	public boolean contieneCacao();
}

class Chocolate implements Comida{
    
    @Override
	public int nutrientes() {
		return 5;
	}

	@Override
	public boolean contieneCacao() {
		return true;
	}
}

abstract class Animal{
	private int hambre;
	
	public void comer(Comida comida) throws Exception{
		if(!this.puedeComer(comida)){
			throw new Exception("No puede comer la comida");
		}
		
		this.hambre -= comida.nutrientes();
	}
		   
	abstract boolean puedeComer(Comida comida);
}

class Gato extends Animal{

    @Override
    public void comer(Comida comida) throws Exception{
        super.comer(comida);
        this.decirMeow();
    }

    @Override
    boolean puedeComer(Comida comida) {
        return !comida.contieneCacao();
    }

    private void decirMeow(){
        System.out.println("Meow");
    }

}